### PR TITLE
expression: wrap arguments when new built-in function

### PR DESCRIPTION
**Pull request by XuHuaiyu**
Types of every arguments and return type should be determined
 when building built-in function signatures,
before this commit, we wrap cast in `runtime` which is not efficient,
this commit wrap cast for args when getFunction.

PTAL @hanfei1991 @shenli

### CODE REVIEW

File: expression/builtin.go [hunk 1]
```diff
@@ -48,18 +60,38 @@ func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc
...
+		return bf, errors.Trace(err)
+	}
 	return baseBuiltinFunc{
 		args:          args,
 		argValues:     make([]types.Datum, len(args)),
 		ctx:           ctx,
 		deterministic: true,
-		tp:            tp,
-	}
+		tp:            retType}, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin.go]: At this line, err must be nil

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @shenli

File: expression/builtin.go [hunk 2]
```diff
@@ -48,18 +60,38 @@ func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc
 	}
 }
 
-// newBaseBuiltinFuncWithTp will be renamed to newBaseBuiltinFunc and replaces the older one later.
-// We'll move the type infer work to expression writer,
-// thus the `tp` attribute is needed for `baseBuiltinFunc`
-// to obtain the FieldType of a ScalarFunction.
-func newBaseBuiltinFuncWithTp(args []Expression, tp *types.FieldType, ctx context.Context) baseBuiltinFunc {
+// newBaseBuiltinFuncWithTp create a built-in function signature with specified types of arguments and return type.
```
Comment (zimulala) [inline expression/builtin.go]: s/create/creates
s/return type/ returns a type

File: expression/builtin.go [hunk 3]
```diff
@@ -48,18 +60,38 @@ func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc
...
+// argTps indicates the types of the args, retType indicates the return type of the built-in function.
+// Every built-in function needs determined argTps and retType when we create it.
+func newBaseBuiltinFuncWithTp(args []Expression, retType *types.FieldType, ctx context.Context, argTps ...argTp) (bf baseBuiltinFunc, err error) {
+	if len(args) != len(argTps) {
+		panic("unexpected length of args and argTps")
```
Comment (zimulala) [inline expression/builtin.go]: Why use `panic`?

File: expression/builtin.go [hunk 4]
```diff
@@ -48,18 +60,38 @@ func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc
...
+			args[i], err = WrapWithCastAsReal(args[i], ctx)
+		case tpDecimal:
+			args[i], err = WrapWithCastAsDecimal(args[i], ctx)
+		case tpString:
+			args[i], err = WrapWithCastAsString(args[i], ctx)
+		case tpTime:
+			args[i], err = WrapWithCastAsTime(args[i], types.NewFieldType(mysql.TypeDatetime), ctx)
+		case tpDuration:
+			args[i], err = WrapWithCastAsDuration(args[i], ctx)
+		}
```
Comment (zimulala) [inline expression/builtin.go]: The `err` value will be covered.

File: expression/expression.go [hunk 5]
```diff
@@ -171,7 +171,16 @@ func evalExprToDecimal(expr Expression, row []types.Datum, sc *variable.Statemen
...
+		res, err = val.ToDecimal(sc)
+		return res, false, errors.Trace(err)
+		// TODO: We maintain two sets of type systems, one for Expression, one for Datum.
+		// So there exists some situations that the two types are not corresponded.
+		// For example, `select 1.1+1.1`
+		// we infer the result type of the sql as `mysql.TypeNewDecimal` which is consistent with MySQL,
+		// but what we actually get is store as float64 in Datum.
+		// So if we wrap `CastDecimalAsInt` upon the result, we'll get <nil> when call `arg.EvalDecimal()`.
+		// This will be fixed after all built-in functions be rewrite correctlly.
+		// return val.GetMysqlDecimal(), false, nil
```
Comment (zimulala) [inline expression/expression.go]: Complete this comment.

File: expression/scalar_function.go [hunk 6]
```diff
@@ -101,10 +102,12 @@ func (sf *ScalarFunction) Clone() Expression {
 	for _, arg := range sf.GetArgs() {
 		newArgs = append(newArgs, arg.Clone())
 	}
-	switch v := sf.Function.(type) {
-	case *builtinCastSig:
-		return NewCastFunc(v.tp, newArgs[0], sf.GetCtx())
-	case *builtinValuesSig:
+	switch sf.FuncName.L {
+	case ast.Cast:
+		newFunc, _ := buildCastFunction(sf.GetArgs()[0], sf.GetType(), sf.GetCtx())
```
Comment (zimulala) [inline expression/scalar_function.go]: Why omit this error?

File: expression/builtin.go (hunk 3 above)
Comment (XuHuaiyu) [inline expression/builtin.go]: len(args) **_must_** equals to len(argTps),
and `argTps` is set by the coder,
so here I use panic to give warning.

File: expression/scalar_function.go (hunk 6 above)
Comment (XuHuaiyu) [inline expression/scalar_function.go]: This is a `Clone` func of ScalarFunction,
the error will never happen.

**Comment by XuHuaiyu** [comment]
PTAL @zimulala

File: expression/builtin_string_test.go [hunk 7]
```diff
@@ -171,7 +171,7 @@ func (s *testEvaluatorSuite) TestConcat(c *C) {
 	}
 	fc := funcs[fcName].(*concatFunctionClass)
 	for _, t := range typeCases {
-		retType := fc.inferType(t.args)
+		retType, _ := fc.inferType(t.args)
```
Comment (zimulala) [inline expression/builtin_string_test.go]: Check this result？

File: expression/builtin_cast.go [hunk 8]
```diff
@@ -249,7 +253,7 @@ func (b *castAsStringFunctionClass) getFunction(args []Expression, ctx context.C
 			sig = &builtinCastStringAsStringSig{bf}
 		}
 	}
-	return sig, errors.Trace(b.verifyArgs(args))
+	return sig.setSelf(sig), errors.Trace(b.verifyArgs(args))
```
Comment (zimulala) [inline expression/builtin_cast.go]: Why do we need to set self-value？
If we really need to do that，line 237 also need to be set.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**